docs/user/openstack/bind-vfio-pci-kernel-module.md

```diff
@@ -0,0 +1,144 @@
+# Binding the vfio-pci kernel module to vhostuser port.
+
+When a VM is attached to a port of type dpdkvhostuserclient, as is the case when vms are
+spun up inside an OpenStack deployment configured for [ovs-dpdk][1] it is possible to use a pmd driver like the [VFIO
+driver][1] on the port.
+Before the interface can be used by a pmd driver,  the vfio-pci kernel module needs to be loaded and bound to
+the desired interface of the vm.
+
+In a regular VM running linux OS, the cli commands would look something like:
+```
+modprobe vfio enable_unsafe_noiommu_mode=1
+modprobe vfio-pci
+echo "0000:00:07.0" > "/sys/bus/pci/drivers/virtio-pci/unbind"
+echo "vfio-pci" > "/sys/bus/pci/devices/0000:00:07.0/driver_override"
+echo "0000:00:07.0" > "/sys/bus/pci/drivers/vfio-pci/bind"```
+```
+
+Entering those commands requires login into the vm and getting adminstrative priviledges.
+
+We can acomplish the same thing when booting a CoreOS vm if we add ta systemd unit similar to
+
+```
+[Unit]
+Description=Vhostuser Interface vfio-pci Bind
+Wants=network-online.target
+After=network-online.target ignition-firstboot-complete.service
+
+[Service]
+Type=oneshot
+EnvironmentFile=/etc/vhostuser-bind.conf
+ExecStart=/usr/local/bin/vhostuser
+
+[Install]
+WantedBy=multi-user.target
+```
+
+The unit will execute the script call "/user/local/bin/vhostuser" son boot, after the network-online  target is reached. See http://systemd.io/ for more details on system units.
+
+Wha is iportant is the contents of _/usr/local/bin/vhostuser_
+
+We could just take the commands we listed above and execute them through the OS cli.
+```
+modprobe vfio enable_unsafe_noiommu_mode=1
+modprobe vfio-pci
+echo "0000:00:07.0" > "/sys/bus/pci/drivers/virtio-pci/unbind"
+echo "vfio-pci" > "/sys/bus/pci/devices/0000:00:07.0/driver_override"
+echo "0000:00:07.0" > "/sys/bus/pci/drivers/vfio-pci/bind"```
+```
+
+That approach does not scale well and there is no way of knowing what is the actual value that should use for the pci
+bus, (instead of "0000:00:07.0" in our example above).
+The script will need know what is the correct pci bus to use based on the network we want to load the kernel
+module for.
+
+This example [script][2] uses the metadata service to find out the mac address of the port attached to the network we
+want to use. The user only needs to provide the openstack network id.
+The script queries the metadata server on boot up. This is the reason we need to reach network-oline target.
```

Comment:
"queries the metadata server on boot up" or read the config drive.

Reply:
`network-online`